src/types/osstr.rs

```diff
@@ -0,0 +1,191 @@
+                py.from_owned_ptr::<PyString>(pystring).into()
+            }
+        }
+
+        #[cfg(windows)]
+        {
+            let wstr: Vec<u16> = std::os::windows::ffi::OsStrExt::encode_wide(self).collect();
+
+            let ptr = wstr.as_ptr();
+            let len = wstr.len() as ffi::Py_ssize_t;
+            unsafe {
+                // HELP: is there some better thing to do here than panic? Some last resort thing to
+                // try?
+                py.from_owned_ptr_or_err::<PyString>(ffi::PyUnicode_FromWideChar(ptr, len))
+                    .expect("OsStr conversion to Python string via FromWideChar failed")
+                    .into()
+            }
+        }
+    }
+}
+
+// It's not possible to implement FromPyObject for OsStr on Windows, because Python's wchar
+// representation can't be directly mapped to OsStr's one
+// HELP: should this implementation be here at all? It may be confusing for PyO3 users when a trait
+// is implemented only on some platforms
+#[cfg(not(windows))]
+impl<'a> FromPyObject<'a> for &'a OsStr {
+    fn extract(ob: &'a PyAny) -> PyResult<Self> {
+        let pystring = <PyString as PyTryFrom>::try_from(ob)?; // Cast PyAny to PyString
+
+        let fs_encoded_bytes: &crate::types::PyBytes = unsafe {
+            ob.py()
+                .from_borrowed_ptr(ffi::PyUnicode_EncodeFSDefault(pystring.as_ptr()))
+        };
+
+        Ok(std::os::unix::ffi::OsStrExt::from_bytes(
+            fs_encoded_bytes.as_bytes(),
+        ))
+    }
+}
+
+impl FromPyObject<'_> for OsString {
+    fn extract(ob: &PyAny) -> PyResult<Self> {
+        #[cfg(not(windows))]
+        {
+            let os_str: &OsStr = ob.extract()?;
+            Ok(os_str.to_os_string())
+        }
+
+        #[cfg(windows)]
+        {
+            let pystring = <PyString as PyTryFrom>::try_from(ob)?; // Cast PyAny to PyString
+
+            // Take the quick and easy shortcut if UTF-8
+            if let Ok(utf8_string) = pystring.to_str() {
+                return Ok(utf8_string.to_owned().into());
+            }
+
+            // The Windows way involves two allocations+copies, unfortunately.
+            // First, we have to convert the opaque PyString into a wide char buffer. Python exposes
+            // no methods to get a view into its internal buffers, so this step can only be done via
+            // copying.
+            // Second, we have to convert the wide char buffer into a OsString. Rust provides no
+            // way to cheaply convert an owned buffer into an OsString; it only exposes a method
+            // involving a copy and an allocation.
+
+            // Get an owned allocated wide char buffer from PyString, which we have to deallocate
+            // ourselves
+            let mut size: ffi::Py_ssize_t = 0;
+            let buffer_ptr =
+                unsafe { ffi::PyUnicode_AsWideCharString(pystring.as_ptr(), &mut size) };
```

Comment:
It could be easier to allocate a `Vec<u16>` of the correct size and use `PyUnicode_AsWideChar` instead.

That would avoid potential issues with `buffer_ptr` being null and also risk of memory leak on panic. (See other comments.)

Reply:
It's somewhat complex, but would look like:
```rust
let size = ffi::PyUnicode_AsWideChar(pystring.as_ptr(), ptr::null_mut(), 0);
if size == -1 {
    return Err(...)
}
let mut buffer = vec![0; size];
assert!(ffi::PyUnicode_AsWideChar(pystring.as_ptr(), buffer.as_mut_ptr(), size), size - 1);

Reply:
I decided to stick with PyUnicode_AsWideCharString and write a `defer_pymem_free` function to handle the deallocation robustly. In my opinion, that way of doing it is simpler and less error-prone than the code snippet by kngwyu

Reply:
Actually, I changed my mind :sweat_smile: The version with a custom allocated Vec is not so bad after all.

However I'm wondering why we can assume that PyUnicode_AsWideChar won't return any error in the second invocation?

Also, if I understand the Python docs correctly, the return value of PyUnicode_AsWideChar is the number of bytes read, not the total number of bytes in the string. If that is true, won't the `size` variable be zero and the whole code blow up?

Reply:
See https://github.com/PyO3/pyo3/pull/1379#issuecomment-758467258 (thanks David) and I found a small mistake in my snippet:
```rust
assert!(ffi::PyUnicode_AsWideChar(pystring.as_ptr(), buffer.as_mut_ptr(), size), size);
```
would be correct.